mdsa/procrustes.py

```diff
@@ -0,0 +1,252 @@
+    # consistent ordering (by sample id) and thus repeatability of results
+
+    # The master order is simply an ascending sort of the sample ids
+    # contained in both the input and reference matrix
+    order = list(set(in_sample_ids) & set(ref_sample_ids))
+
+    if len(order) == 0:
+        raise ValueError(
+            'No overlapping sample IDs in the two OrdinationResults '
+            'matrices.')
+    elif len(order) != len(in_sample_ids) or len(order) != len(ref_sample_ids):
+        warn('Procrustes: some sample ids of input_matrix do not overlap with '
+             'sample ids of reference_matrix. Non-overlapping corresponding '
+             'samples in matrices will be dropped for procrustes procedure.',
+             RuntimeWarning
+             )
+    in_matrix = _reorder_rows(in_matrix, in_sample_ids, order)
+    ref_matrix = _reorder_rows(ref_matrix, ref_sample_ids, order)
+
+    # If randomization function given, use it to shuffle the reference matrix
+    # (used for monte carlo random trial)
+    if randomize_fn:
+        return_matrix = randomize_fn(ref_matrix)
+
+        # if randomize_fn did not shuffle in-place, then update ref_matrix
+        # with the shuffled return_matrix manually
+        if return_matrix is not None:
+            ref_matrix = return_matrix
+
+    # Apply padding to matrices to ensure equal dimensionality
+    in_matrix, ref_matrix = _pad_matrices(in_matrix, ref_matrix)
+
+    # Run the Procrustes analysis
+    in_matrix_transformed, ref_matrix_transformed, m_squared = \
+        scipy_procrustes(ref_matrix, in_matrix)
+    return m_squared
+
+
+def calc_p_value(trials, count_better):
+    """
+    Calculate p-value and truncate to appropriate number of significant digits
+    (significant digits scaled logarithmically to number of trials).
+
+    Parameters
+    ----------
+    trials : int
+        Number of monte-carlo trials to run, 1000 by default.
+    count_better : int
+        Number of trials that resulted in a lower m^2 (i.e. better) than
+        m_squared_actual.
+
+    Returns
+    -------
+    p_value : float
+        The number of trials that resulted in a lower, i.e. better, m^2
+        divided by the total number of trials. P-Value is truncated to
+        a reasonable number of significant figures (scaled by log10 to
+        the number of trials).
+    """
+    pval = float(count_better) / float(trials)
+    decimal_places = int(np.log10(trials + 1))
+    return float(('%1.' + '%df' % decimal_places) % pval)
+
+
+def _reorder_rows(matrix, sample_ids, order):
+    """
+    Arrange the rows in matrix to correspond to order
+
+    Note: order is the master list here -- if a sample id is
+    not included in order, that coord will not be included in
+    the results. All sample ids in order however must be in
+    sample_ids.
+    """
+    result = []
+    for sample_id in order:
+        try:
+            result.append(matrix[sample_ids.index(sample_id)])
```

Comment:
This is a list look up (`O(n)`) and I think only needs to be done once. What about doing the sample ID check explicitly outside of the for loop and use a lookup structure in the loop (e.g., `lookup = {i: idx for idx, i in enumerate(order)`)?

Reply:
Yes you're right, a lookup structure would be better to make this more generically performant. 

But I presumed there are no duplicate sample ids in the list `sample_ids` (which originates from OrdinationResults). If that is true, then there should never be repeated hits to `index(sample_id)` with the same `sample_id`,  obviating the need for a lookup function.

Am I correct in this reasoning or missing something?

Reply:
Isn't the call to `sample_ids.index` made `order` times? What about the following:

```python
if set(order) != set(sample(ids)):
    raise ValueError('Procrustes row reordering etc etc...')

lookup = {i: idx for idx, i in enumerate(order)}
for sample_id in order:
    result.append(matrix[lookup[sample_id]])
```

Reply:
Got it, thanks!

** slight correction to param passed to enumerate: `lookup = {sample_id: idx for idx, sample_id in enumerate(sample_ids)}`